gitpod-io/gitpod
# [usernamespaces] Introduce custom seccomp profile

This PR introduces a custom seccomp profile based on containerd's runtime default.
We do this so that user-namespaced workspaces don't need to use `unconfined`, but use a sensible profile, too.

The custom profile is basically the `runtime/default` profile of containerd + `mount`, `chroot` and `clone`.

### How does it work?
1. we [produce our own seccomp profile](https://github.com/gitpod-io/gitpod/blob/1c0df61d287a6e34bdf565411068aaecc65675d4/components/ws-daemon/seccomp-profile-installer/main.go#L25-L33) using containerd's runtime default. We use the list of allowed capabilities from a ws-manager test fixture as input for that profile (containerd's default seccomp profile explicitly allows syscalls if the pod gets the necessary capabilities). We might find that this approach is too restrictive, given the additional capabilities our users get in the user namespace.
2. we copy the seccomp profile to the node using a new init container of `ws-daemon`

The seccomp profiles carry the Gitpod version in their name, to facilitate different versions on the same node (e.g. in devstaging, but also during prod deployments).

### How to test?
- start a workspace, inspect the pod `kubectl get pod -o yaml -l component=workspace`, ensure that it's using the `runtime/default` seccomp profile (`seccomp.security.alpha.kubernetes.io/pod` annotation)
- set the user namespace feature flag: `leeway run components/ws-daemon:enabled-userns-ff`
- start another workspace
  - observe that it actually starts
  - ensure that it's using a custom seccomp profile (`seccomp.security.alpha.kubernetes.io/pod` annotation), e.g. `localhost/workspace_default_...json`
